### ISSUE TITLE
[P1] How could we apply Zero-knowledge proof(0KP) on our service?

### ISSUE THREAD

**Issue by seiwonpark** [kookmin-sw/capstone-2023-40#91]
Regarding [THE KNOWLEDGE COMPLEXITY OF INTERACTIVE PROOF SYSTEMS](http://people.csail.mit.edu/silvio/Selected%20Scientific%20Papers/Proof%20Systems/The_Knowledge_Complexity_Of_Interactive_Proof_Systems.pdf)

**Comment by seiwonpark** [comment]
Referring to [THE KNOWLEDGE COMPLEXITY OF INTERACTIVE PROOF SYSTEMS](http://people.csail.mit.edu/silvio/Selected%20Scientific%20Papers/Proof%20Systems/The_Knowledge_Complexity_Of_Interactive_Proof_Systems.pdf), 

> Although we find the idea of a zero-knowledge interactive proof system fascination in iteself, the main motivation for it and the main applications of it are in the area of cryptographic protocols.

this could be useful on our service by adding zero-knowledge protocol to our filter chain.

**Comment by seiwonpark** [comment]
And there's another interesting paper [ZeroTwo](https://netsec.ethz.ch/publications/papers/ZeroTwo.pdf).
(~~Not the.....name of dance~~)

**Comment by bezzang-dev** [comment]
From what I understand.. When the survey participant completes the survey, the server must encrypt the response and come to the server, which saves the encrypted response to the db. 
And later, when a client requests a challenge from the server to prove that the client has participated in the survey, the server compares it with the response information in the saved db.

I don't know if it's right because the theory is difficult...

I think it's similar to _Blind signatures_.

**Anyway, how do we implement it!?**

**Comment by seiwonpark** [comment]
Why do you think "the surver must encrypt the response"? That's different from what I've understood. Could you explain it why?

**Comment by seiwonpark** [comment]
"Anyway, how do we implement it!?" >> By adding filter since it's the core authentication middleware which intercepts requests from client.

**Comment by bezzang-dev** [comment]
Oh, I correct "the surver must encrypt the response" to "the client has to send request encrypted.".

**Comment by seiwonpark** [comment]
That still doesn't make sense. So you mean after answer all the questions in a survey, encrypt the whole data and send it to server?

**Comment by bezzang-dev** [comment]
I understood it like that, but now I think it's wrong.

Let me explain what I understand conceptually.
For example,
The client said, "I submitted a response to the survey with the necessary authentication, but I don't want to show the server the personal information used for my authentication."
In this situation, we need to apply 0KP.
0KP allows a client to prove the authenticity of responding the survey without revealing any additional information.

So, How 0KP works is... I'll study more.

**Comment by seiwonpark** [comment]
Well that's very different from mine. What I've understood is,

not actually hiding("I don't want my information to be shown to server used for authentication") but it's just about authentication. My understanding is 0KP is just another kind of authentication method like token or session. So we could replace all our session authentication totally with zero-knowledge proof system or combine them together or with token. So it doesn't have to know "personal information" or "client's request body" or the context as well. Since it'll be handled on every filter chain on every single request.

And "0KP allows a client to prove the authenticity of responding the survey without revealing any additional information." is just repeating the definition of 0KP and it does nothign. And this is acutally what I asked you "How?". My answer was adding another filter. How about your idea?

**Comment by sanghyun1128** [comment]
I understood that, when submitting a survey, we submit excluding user information to ensure anonymity, and that could lead to duplicate participation, so we should use 0kp for that, is that right...? 
I don't know if it's meaningful because it's managed by the same server as user information or surveys anyway

**Comment by seiwonpark** [comment]
Well then I don't know how to save data for User and Survey entity because I don't think they should remain related. Let me just ask this to professor.

**Comment by bezzang-dev** [comment]
I thought "Shouldn't there be some logic implementation that validates any encrypted 0KP information sent by the client on the server??", so I was wondering how to do it...
And I don't have any ideas yet. ; (

but Looking at @sanghyun1128 comment , it seems that the intention of why to use 0KP is ambiguous...

**Comment by seiwonpark** [comment]
Yeah same idea. Even [ZeroTwo](https://netsec.ethz.ch/publications/papers/ZeroTwo.pdf) says
> There are several reasons for this: key management is difficult; users need a trusted device with storage, communication, and computation capabilities; and the authentication process must be fast and convenient. We believe, however, that the democratization of smartphones with embedded biometric sensors, unlimited cellular data plans, and new communication standards such as WebRTC now make asymmetric protocols a viable option for user authentication.

But still, I have no idea where to store data from client. This arises the same question for JWT token (http-only cookie, is that the real answer for security?).
